orchestrator/services/processes.py

```diff
@@ -155,54 +138,102 @@ def _db_log_step(
             else:
                 p.assignee = Assignee.SYSTEM
 
-        # check if last error state is identical to determine if we add a new step or update the last one
-        last_db_step = p.steps[-1] if len(p.steps) else None
+    else:
+        p.failed_reason = None
+        p.traceback = None
+
+    return p
+
 
+def _get_current_step_to_update(
+    stat: ProcessStat, p: ProcessTable, step: Step, process_state: WFProcess
+) -> ProcessStepTable:
+    """Checks if last error state is identical to determine if we add a new step or update the last one."""
+    step_state: State = process_state.unwrap()
+    current_step = None
+    last_db_step = p.steps[-1] if len(p.steps) else None
+    current_sub_step, current_step_group = step_state.get("__sub_step"), step_state.get("__step_group")
+    db_sub_step, db_step_group = (
+        (last_db_step.state.get("__sub_step"), last_db_step.state.get("__step_group")) if last_db_step else (None, None)
+    )
+
+    if current_sub_step is None and db_sub_step is None:
+        pass  # Skip step group logic
+    elif (db_sub_step is None and current_sub_step) or (db_sub_step and current_sub_step is None):
+        # New entry
+        current_step = ProcessStepTable(
+            process_id=stat.process_id,
+            name=current_step_group or db_step_group,
+            status=process_state.status,
+            state=step_state,
+            created_by=stat.current_user,
+        )
+    else:
+        # Update last step entry in db
+        logger.info("Will be merged")
+        last_db_step.status = process_state.status
+        last_db_step.state = step_state
+        current_step = last_db_step
```

Comment:
could change the first `current_step = None` to `current_step = last_db_step` and remove it here

Reply:
I find this a bit more intuitive.

Reply:
Ideally, we don't overwrite the value of local variables ever. But I think this is better done as part of a larger refactor of this module. The current_step is again used further down the line and checked against None.
Note that I copied this part from the previous implementation and added an if branch for sub_steps. I'd like to refactor the whole function, but again, better for a separate issue.

Reply:
Thought it was such a small change that it didn't need to wait but if you want to refactor the whole function it can wait.